File: apps/backends/bun/src/services/checkout.ts
import postgres from 'postgres';
import { CheckoutRequest, CheckoutResponse, OrderItem } from '../types/index';
import { getProduct, insertOrder } from '../db/queries';
import { computeFraudScore } from './fraud';
import { calculateTax } from '../clients/tax';

export async function processCheckout(
  req: CheckoutRequest,
  pool: postgres.Sql,
  taxServiceUrl: string,
): Promise<CheckoutResponse> {
  if (!req.items || req.items.length === 0 || req.items.length > 8) {
    throw new Error('Cart must have 1-8 items');
  }

  if (!isValidUUID(req.customer_id)) {
    throw new Error('Invalid customer ID');
  }

  let subtotal = 0;
  const orderItems: OrderItem[] = [];

  for (const item of req.items) {
    if (!isValidUUID(item.product_id)) {
      throw new Error('Invalid product ID');
    }

    const product = await getProduct(pool, item.product_id);
    if (product.stock < item.quantity) {
      throw new Error('Insufficient stock');
    }

    subtotal += product.price_cents * item.quantity;
    orderItems.push({
      product_id: item.product_id,
      quantity: item.quantity,
      price_cents: product.price_cents,
    });
  }

  const taxResp = await calculateTax(taxServiceUrl, subtotal, req.state);
  const fraudScore = computeFraudScore(subtotal, orderItems.length);
  const orderId = crypto.randomUUID();
  const total = subtotal + taxResp.tax_cents;

  await insertOrder(pool, orderId, req.customer_id, total, taxResp.tax_cents, orderItems);

  return {
    order_id: orderId,
    total_cents: total,
    tax_cents: taxResp.tax_cents,
    fraud_score: fraudScore,
  };
}

function isValidUUID(uuid: string): boolean {
  const uuidRegex = /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i;
  return uuidRegex.test(uuid);
}

File: apps/backends/rust/src/domain/checkout.rs
use uuid::Uuid;
use serde::{Deserialize, Serialize};
use sqlx::PgPool;
use crate::db::orders;
use crate::clients::tax::TaxClient;
use crate::domain::fraud::compute_fraud_score;

#[derive(Deserialize)]
pub struct CheckoutRequest {
    pub customer_id: String,
    pub items: Vec<CartItem>,
    pub state: String,
}

#[derive(Deserialize)]
pub struct CartItem {
    pub product_id: String,
    pub quantity: i32,
}

#[derive(Serialize)]
pub struct CheckoutResponse {
    pub order_id: String,
    pub total_cents: i32,
    pub tax_cents: i32,
    pub fraud_score: i32,
}

pub async fn process_checkout(
    req: CheckoutRequest,
    pool: &PgPool,
    tax_client: &TaxClient,
) -> Result<CheckoutResponse, Box<dyn std::error::Error>> {
    if req.items.is_empty() || req.items.len() > 8 {
        return Err("Cart must have 1-8 items".into());
    }
    
    let customer_id = Uuid::parse_str(&req.customer_id).map_err(|_| "Invalid customer ID")?;
    let mut subtotal = 0i32;
    let mut order_items = Vec::new();
    
    for item_req in req.items {
        let product_id = Uuid::parse_str(&item_req.product_id).map_err(|_| "Invalid product ID")?;
        let product = orders::get_product(pool, product_id).await.map_err(|_| "Product not found")?;
        
        if product.stock < item_req.quantity {
            return Err("Insufficient stock".into());
        }
        
        subtotal += product.price * item_req.quantity;
        order_items.push(orders::OrderItem {
            product_id,
            quantity: item_req.quantity,
            price: product.price,
        });
    }
    
    let tax_resp = tax_client.calculate_tax(subtotal, &req.state).await?;
    let fraud_score = compute_fraud_score(subtotal, order_items.len());
    let order_id = Uuid::new_v4();
    let total = subtotal + tax_resp.tax_cents;
    
    orders::insert_order(
        pool,
        orders::Order { id: order_id, customer_id, total_cents: total, tax_cents: tax_resp.tax_cents },
        order_items,
    ).await?;
    
    Ok(CheckoutResponse {
        order_id: order_id.to_string(),
        total_cents: total,
        tax_cents: tax_resp.tax_cents,
        fraud_score,
    })
}
